src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs

```diff
@@ -10,24 +10,21 @@ namespace NuGet.Packaging
 {
     public static class TopologicalSortUtility
     {
+        private static readonly PackageInfoComparer DefaultComparer = new PackageInfoComparer();
         /// <summary>
         /// Order dependencies by children first.
         /// </summary>
         public static IReadOnlyList<PackageDependencyInfo> SortPackagesByDependencyOrder(
             IEnumerable<PackageDependencyInfo> packages)
         {
-            var sorted = new List<PackageDependencyInfo>();
-            var toSort = packages.Distinct().ToList();
+            var toSort = packages.Distinct().Select(package => new PackageInfo(package)).ToArray();
```

Comment:
Looking at in a more traditional profiler, you can see the top level function `SortPackagesByDependencyOrder` allocations are reduced, and the subfunction `GetParentCount` allocations are eliminated:

![](https://aoa.blob.core.windows.net/aspnet/top_allocs.png)

Looking at execution time; you can see most of the time is spent in the subfunction `GetParentCount` and that is dramatically reduced; while even the top level function's self time is reduced:

![](https://aoa.blob.core.windows.net/aspnet/top_time.png)